server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -30,21 +43,80 @@ class ObjectService(googleStorageService: GoogleStorageService[IO]) extends Http
+      res <- storageLink.fold[IO[MetadataResponse]](IO.raiseError(NotFoundException(s"No storage link found for ${req.localObjectPath}"))) {
+        sl =>
+          for {
+            bucketAndObject <- IO.fromEither(JsonCodec.parseGsDirectory(sl.cloudStorageDirectory.renderString).leftMap(e => InternalException(e)))
+            metadata <- googleStorageService.getObjectMetadata(bucketAndObject.bucketName, bucketAndObject.blobName, Some(traceId)).compile.lastOrError
+            result <- metadata match {
+                        case google2.GetMetadataResponse.NotFound => IO.raiseError(NotFoundException(s"Object(${bucketAndObject}) not found in Google storage"))
+                        case google2.GetMetadataResponse.Metadata(crc32, userDefinedMetadata) =>
+                          for {
+                            isLinkedString <- IO.fromEither(userDefinedMetadata.get("isLinked").toRight(NotFoundException("isLinked key missing from object metadata ")))
+                            isLinked <- IO(isLinkedString.toBoolean)
+                            fileBody <- googleStorageService.getObject(bucketAndObject.bucketName, bucketAndObject.blobName, Some(traceId)).compile.to[Array]
+                            calculatedCrc32 = Crc32c.calculateCrc32c(fileBody)
+                            syncStatus = if (fileBody.length == 0) SyncStatus.DeletedRemote
+                              else if (calculatedCrc32 == crc32) SyncStatus.Live
+                              else SyncStatus.Desynchronized
+                            lastEditedBy <- IO.fromEither(userDefinedMetadata.get("lastEditedBy").map(WorkbenchEmail).toRight(NotFoundException("lastEditedBy key missing from object metadata ")))
+                            lastEditedTimeString <- IO.fromEither(userDefinedMetadata.get("lastEditedTime").toRight(NotFoundException("lastEditedTime key missing from object metadata ")))
+                            lastEditedTime <- IO(Instant.ofEpochMilli(lastEditedBy.value.toLong))
+                          } yield MetadataResponse(isLinked, syncStatus, lastEditedBy, lastEditedTime, sl.cloudStorageDirectory, true, sl) //TODO: fix isExecutionMode
+                      }
+          } yield result
+      }
+    } yield res
 
-  def safeDelocalize(req: SafeDelocalize): IO[Unit] = ???
+  def safeDelocalize(req: SafeDelocalize): IO[Unit] = {
+    for {
+      traceId <- IO(TraceId(randomUUID()))
+      storageLinks <- readJsonFileToA[IO, List[StorageLink]](pathToStorageLinks).compile.lastOrError
+      storageLink = storageLinks.find(_.localBaseDirectory == req.localObjectPath)
+      metadata <- checkMetadata(GetMetadataRequest(req.localObjectPath))
+      _ <- if(metadata.syncStatus == SyncStatus.Live) {
+        io.file.readAll[IO](Paths.get(req.localObjectPath.asString), blockingEc, 4096).compile.to[Array].flatMap {
+          body =>
+            googleStorageService.storeObject(GcsBucketName(""), GcsBlobName(""), body, "text/plain", Some(traceId))
```

Comment:
I'm curious - where happens to this traceId?

Reply:
it'll show up in logs. Mostly the intention is we'll be able to find all logs related say one request. But later I realize it probably doesn't help that much because welder lives on user's vm